pkg/sql/sem/builtins/builtins.go

```diff
@@ -94,53 +96,83 @@ func categorizeType(t types.T) string {
 
 var digitNames = []string{"zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"}
 
+// BuiltinDefinition represents a built-in function before it becomes
+// a tree.FunctionDefinition.
+type BuiltinDefinition struct {
+	tree.FunctionProperties
+	Overloads []tree.OverloadDefinition
+}
+
+// defProps is used below to define built-in functions with default properties.
+func defProps() tree.FunctionProperties { return tree.FunctionProperties{} }
+
+// arrayProps is used below for array functions.
+func arrayProps() tree.FunctionProperties { return tree.FunctionProperties{Category: categoryArray} }
+
+// arrayPropsNullableArgs is used below for array functions that accept NULLs as arguments.
+func arrayPropsNullableArgs() tree.FunctionProperties {
+	p := arrayProps()
+	p.NullableArgs = true
+	return p
+}
+
+func makeBuiltin(
+	props tree.FunctionProperties, overloads ...tree.OverloadDefinition,
+) BuiltinDefinition {
+	return BuiltinDefinition{
+		FunctionProperties: props,
+		Overloads:          overloads,
+	}
+}
+
 // Builtins contains the built-in functions indexed by name.
-var Builtins = map[string][]tree.Builtin{
+var Builtins = map[string]BuiltinDefinition{
```

Comment:
Can you make this a `*BuiltinDefinition` instead? Otherwise it forces the code that accesses the map to make a copy of the struct, since you can't get a pointer to a map value.